### PR TITLE
[H2] Support H2 service/operations

### PR DESCRIPTION
**Pull request by cjyclaire**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

This PR is aimed for async usage for operations over HTTP2 protocol (bi-direction/uni-direction)

~~CI currently failed as expected~~
~~[WIP]~~

~~TODOs~~
1. ~~event signal param validation~~
2. ~~retry story~~
3. ~~code gem docs for async operations/shapes/usage~~
4. ~~dependency story~~
5. ~~CI~~

### CODE REVIEW

**Comment by cjyclaire** [comment]
this PR is now ready for review :)

File: gems/aws-sdk-core/lib/aws-sdk-core/binary/decode_handler.rb [hunk 1]
```diff
@@ -23,11 +23,27 @@ def attach_eventstream_listeners(context, rules)
...
             rules,
+            context.operation.output,
+            context.operation.errors,
             context.http_response.body,
-            context[:event_stream_handler])
+            output_handler)
+          if input_emitter = context[:input_event_emitter]
+            # events need to be send in order
+            # every event signature requries prior signature
+            thread = Thread.new do
```
Comment (awood45) [inline gems/aws-sdk-core/lib/aws-sdk-core/binary/decode_handler.rb]: This looks like we are leaking threads. How are we cleaning these up?

File: gems/aws-sdk-core/lib/aws-sdk-core/binary/encode_handler.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+module Aws
+  module Binary
+
+    # @api private
+    class EncodeHandler < Seahorse::Client::Handler
+
+      def call(context)
+        if eventstream_member = eventstream_input?(context)
+          # TODO, remove handler? not need? #delete
```
Comment (awood45) [inline gems/aws-sdk-core/lib/aws-sdk-core/binary/encode_handler.rb]: Can we resolve this TODO question?

File: gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb [hunk 3]
```diff
@@ -0,0 +1,196 @@
+if RUBY_VERSION >= '2.3'
+  require 'http/2'
```
Comment (awood45) [inline gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb]: And if it's an earlier Ruby version? We need to have a user-friendly error when client creation is attempted in an older version of Ruby.

File: gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  module Client
+    # @api private
+    module H2
+
+      # H2 Connection build on top of `http/2` gem
+      # with TLS layer plus ALPN, requires:
+      # Ruby >= 2.3 and OpenSSL >= 1.0.2
+      class Connection
+
+        OPTIONS = {
```
Comment (awood45) [inline gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb]: How are we surfacing these defaults to user-facing documentation?

File: gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            debug_output("opened")
+
+            @socket = OpenSSL::SSL::SSLSocket.new(tcp, _tls_context)
+            @socket.sync_close = true
+            @socket.hostname = endpoint.host
+
+            debug_output("starting TLS for #{endpoint.host}:#{endpoint.port} ...")
+            @socket.connect
+            debug_output("TLS over ALPN established")
+            # TODO, confirm with Transcribe
```
Comment (awood45) [inline gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb]: Let's resolve this TODO item.

File: gems/aws-sdk-core/lib/seahorse/client/plugins/h2.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+require 'seahorse/client/h2/handler'
+
+module Seahorse
+  module Client
+    module Plugins
+      class H2 < Plugin
+
+        # H2 Client
+        option(:max_concurrent_streams, default: 100, doc_type: Integer, docstring: '')
```
Comment (awood45) [inline gems/aws-sdk-core/lib/seahorse/client/plugins/h2.rb]: Need these docstrings filled out. Should answer my question from above, though.

File: gems/aws-sdk-core/spec/seahorse/client/async_base_spec.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
+require_relative '../../spec_helper'
+
+module Seahorse
+  module Client
+    describe AsyncBase do
+      if RUBY_VERSION >= '2.3'
```
Comment (awood45) [inline gems/aws-sdk-core/spec/seahorse/client/async_base_spec.rb]: There needs to be a counter to this, where the RUBY_VERSION is < 2.3 and we raise an appropriate error.

File: gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb (hunk 4 above)
Comment (cjyclaire) [reply gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb]: Addressed in the plugin file doc_strings :)

File: build_tools/aws-sdk-code-generator/lib/aws-sdk-code-generator/client_operation_list.rb [hunk 8]
```diff
@@ -42,8 +67,12 @@ def initialize(options)
         @documentation = options.fetch(:documentation)
         @streaming = options.fetch(:streaming)
         @eventstream_output = !!options.fetch(:eventstream_output)
-        @eventstream_member = @eventstream_output ?
+        @eventstream_input = !!options.fetch(:eventstream_input)
+        @output_eventstream_member = @eventstream_output ?
           options.fetch(:eventstream_output) : nil
+        @input_eventstream_member = @eventstream_input ?
+          options.fetch(:eventstream_input) : nil
+        @bidirectional = @eventstream_output && @eventstream_input
```
[the remark this answers is not in the document]
Comment (cjyclaire) [reply build_tools/aws-sdk-code-generator/lib/aws-sdk-code-generator/client_operation_list.rb]: Yep, it's affecting docs and `Aws::Svc::EventStreams` class generation,
samples can be viewed at: https://github.com/aws/aws-sdk-ruby/pull/1945/files#diff-78e820efc317198d8b9bab2f79d7fa47

File: gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb (hunk 3 above)
Comment (cjyclaire) [reply gems/aws-sdk-core/lib/seahorse/client/h2/connection.rb]: Yep, sample: https://github.com/aws/aws-sdk-ruby/pull/1945/files#diff-02d74b79db5eac02ca88a5b6f2154efb
(the diff is hidden by Github :()

Basic, I avoid to throw build/load errors but errors out immediately when they are trying to initialize an AsyncClient

File: gems/aws-sdk-core/aws-sdk-core.gemspec [hunk 9]
```diff
@@ -13,8 +13,9 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency('jmespath', '~> 1.0')
   spec.add_dependency('aws-partitions', '~> 1.0')
-  spec.add_dependency('aws-sigv4', '~> 1.0') # necessary for making Aws::STS API calls
-  spec.add_dependency('aws-eventstream', '~> 1.0') # necessary for binary eventstream
+  spec.add_dependency('aws-sigv4', '~> 1.1') # necessary for making Aws::STS API calls
+  spec.add_dependency('aws-eventstream', '~> 1.0', '>= 1.0.2') # necessary for binary eventstream
+  spec.add_dependency('http-2', '~> 0') # necessary for making h2 calls
```
[the remark this answers is not in the document]
Comment (cjyclaire) [reply gems/aws-sdk-core/aws-sdk-core.gemspec]: hmm makes sense, currently 
https://github.com/igrigorik/http-2/releases

maybe 
```
spec.add_dependency('http-2', '~> 0.10')
```
?

**Comment by cjyclaire** [comment]
Ready for another round of review 🎉

**Pull request merged**